### ISSUE TITLE
train_dataset.create_hdf5_dataset(resize=(width,height),..) does not scale ground truth coordinates 

### ISSUE THREAD

**Issue by svebert** [pierluigiferrari/ssd_keras#138]
Hi,
I am _playing_ with your nice SSD implementation (with the ultimative goal to include mobilenet as base net and to have everything compatible for inference with OpenCV DNN in C++)

I just started with the ssd7_training example, but with my own training data/labels. Therefore, I changed 
`train_dataset.create_hdf5_dataset(resize=False,...) `
to
`train_dataset.create_hdf5_dataset(resize=(224,224),...))`
resulting in weird bounding box coordinates

The reason is, that the training images will be resized, but the ground truth coordinates will not. This is a bug, I guess.

I fixed it with this code

```.py
		for i in tr:

			# Store the image.
			with Image.open(self.filenames[i]) as image:

				image = np.asarray(image, dtype=np.uint8)

				# Make sure all images end up having three channels.
				if image.ndim == 2:
					image = np.stack([image] * 3, axis=-1)
				elif image.ndim == 3:
					if image.shape[2] == 1:
						image = np.concatenate([image] * 3, axis=-1)
					elif image.shape[2] == 4:
						image = image[:,:,:3]

				if resize:
					sizeRaw = image.shape					
					image = cv2.resize(image, dsize=(resize[1], resize[0]))
					dSizeFactorH = sizeRaw[0]/(image.shape[0]*1.0)
					dSizeFactorW = sizeRaw[1]/(image.shape[0]*1.0)
				# Flatten the image array and write it to the images dataset.
				hdf5_images[i] = image.reshape(-1)
				# Write the image's shape to the image shapes dataset.
				hdf5_image_shapes[i] = image.shape

			# Store the ground truth if we have any.
			if not (self.labels is None):
				#rescale the ground thruth box coordinates if images were rescaled
				if resize:
					self.labels[i][0][self.labels_output_format.index('xmax')]=int(self.labels[i][0][self.labels_output_format.index('xmax')]/dSizeFactorW)
					self.labels[i][0][self.labels_output_format.index('xmin')]=int(self.labels[i][0][self.labels_output_format.index('xmin')]/dSizeFactorW)
					self.labels[i][0][self.labels_output_format.index('ymax')]=int(self.labels[i][0][self.labels_output_format.index('ymax')]/dSizeFactorH)
					self.labels[i][0][self.labels_output_format.index('ymin')]=int(self.labels[i][0][self.labels_output_format.index('ymin')]/dSizeFactorH)
				labels = np.asarray(self.labels[i])
				# Flatten the labels array and write it to the labels dataset.
				hdf5_labels[i] = labels.reshape(-1)
				# Write the labels' shape to the label shapes dataset.
				hdf5_label_shapes[i] = labels.shape

```

**Comment by MBoaretto25** [comment]
Were you able to use the ssd7 model in opencv DNN? I'm thinking on using opencv as backend, however i'm worried about the complexity of converting the keras layers due to compatibility issues in opencv.

**Comment by MBoaretto25** [comment]
Regarding your issue with the bouding box, I faced the same problem. So what I've done is to rescale coordinates in the csv parser.

**Comment by svebert** [comment]
Hi Marco,

yes I was able to use ssd models in opencv dnn. But you are right, it was hard work to convert it from keras. Also I had to implement the last ssd stages (the parts which are not directly based on convolutional networks) in C++ (port it from python to C++).
Actually training the network with keras is only half of the full work. The other half is porting the resulting network to opencv and implement all in C++.

Mit freundlichen Grüßen / Best regards

Sven Hans
DERMALOG Identification Systems GmbH
Systems Engineer

Mittelweg 120 | 20148 Hamburg
Fon +49 40 413227 0

Fax +49-40-413227-40

Mail Sven.Hans@dermalog.com<mailto:Sven.Hans@dermalog.com>

Web www.dermalog.com<https://www.dermalog.com>


This e-mail may contain information that is confidential or proprietary to DERMALOG Identification Systems GmbH. Any unauthorized disclosure, distribution or other use is prohibited.

P Please consider the environment before printing this email

**Comment by MBoaretto25** [comment]
Thanks for the answer @svebert. I'll try to implement a compatible version of ssd7, if I have any sucess I will post something here.

**Comment by MBoaretto25** [comment]
@svebert how did you handle the AnchorBox convertion to the opencv model, I'm having a lot of trouble trying to conver the np.array into a tensor output for the layer.

**Comment by s-g-stavely** [comment]
This is still a problem. I'd like to note that the fix in svebert's comment works, but it assumes that there is only one labelled bounding box per image, so if you've got more boxes you'll need to modify the use of `self.labels[i][0]` to iterate over all the boxes.
